File: 2022/04/02.py
import sys


def overlaps(a, b, x, y):
    return a <= y and b >= x or x <= b and y >= a


def solve(line):
    elf1, elf2 = line.split(',')
    a, b = [int(x) for x in elf1.split('-')]
    x, y = [int(x) for x in elf2.split('-')]
    return 1 if overlaps(a, b, x, y) else 0


def main():
    line = sys.stdin.readline()
    result = 0

    while line:
        result += solve(line)
        line = sys.stdin.readline()

    print(result)


main()

File: 2024/src/01/02.ts
import * as path from 'path';
import { readFile } from '../utils/input';
import { pj } from '../utils/print';

const sampleFilepath = path.join(__dirname, 'sample');
const inputFilepath = path.join(__dirname, 'input');

const main = () => {
        // const rawInput = readFile(sampleFilepath);
        const rawInput = readFile(inputFilepath);
        // pj(rawInput);

        const input = parseInput(rawInput);
        // pj(input);

        const result = solve(input);
        pj(result);
};

const parseInput = (rawInput: string) => {
        const matrix: number[][] = [[], []];

        for (const line of rawInput.split('\n')) {
                const values = line.split(/\s+/);
                for (let i = 0; i < values.length; i++) {
                        matrix[i].push(Number(values[i]));
                }
        }

        return matrix;
};

const solve = (input: number[][]) => {
        // 1. sort
        for (let i = 0; i < input.length; i++) {
                input[i] = input[i].sort((a, b) => a - b);
        }

        // 2. build "occurrences" map
        const occurrenceTimesMap: { [key: number]: number } = {};
        for (const number of input[1]) {
                if (!occurrenceTimesMap[number]) occurrenceTimesMap[number] = 0;
                occurrenceTimesMap[number] += 1;
        }

        // 3. sum for each number
        let total = 0;
        for (let i = 0; i < input[0].length; i++) {
                const number = input[0][i];
                const occurrenceTimes = occurrenceTimesMap[number] || 0;
                total += number * occurrenceTimes;
        }

        // 4. return total sum
        return total;
};

main();

export {};
